### ISSUE TITLE
Potential ambiguities in clustering of lesions 

### ISSUE THREAD

**Issue by fedorov** [pylidc/pylidc#18]
@notmatthancock I was (again) reading this paper, and came across the example shown below (Fig.8):

Armato et al. The Lung Image Database Consortium (LIDC) and Image Database Resource Initiative (IDRI): a completed reference database of lung nodules on CT scans. Med. Phys. 38, 915–931 (2011). https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3041807/

![image](https://user-images.githubusercontent.com/313942/48513682-67a5f500-e82a-11e8-8a10-67ba93d10300.png)

What is expected from the clustering of annotations into nodules, as implemented in pylidc, in this situation? I do not know which case was used in that illustration, so cannot quickly check what is actually happening.

**Comment by fedorov** [comment]
This is very helpful. I should be able to quickly check those tomorrow. Thanks!

**Comment by fedorov** [comment]
55 is a good example case shown below (I reformatted the slice to show oblique plane, since otherwise  there is no single slice showing lesion annotated in pieces and as a whole).

Two sub-groups of nodules:

![image](https://user-images.githubusercontent.com/313942/48571852-e86dfb00-e8d5-11e8-8f3d-ddf9471b99ae.png)

Single nodule:

![image](https://user-images.githubusercontent.com/313942/48571887-fc196180-e8d5-11e8-9352-78b1fc43ef3b.png)

I believe saving those components as independent nodules is fine, since due to the way the study was set up, we cannot determine which of the sub-nodules belong to which readers. And even if we knew that, we would not know if the reader was annotating the same nodule, or interpreted that as different nodules, since annotations do not have nodule identifiers.

I think what pylidc is doing makes sense. I will examine other cases visually as well.

For the sake of completeness, here is the list of cases (generated with [this script](https://github.com/QIICR/lidc2dicom/blob/master/checkClusters.py)) where either warnings were raised, or there was a mismatch between the total number of annotations and the number of annotations included in pylidc-defined annotations clusters:

* 55 - warning
* 92 - warning
* 132 - 17 annotations unaccounted for
* 137 - warning
* 152- warning
* 204 - warning
* 252 - warning
* 315 - 19 annotations unaccounted for
* 332 - 9 annotations unaccounted for
* 340 - warning
* 355 - 3 annotations unaccounted for
* 365 - 1 annotation unaccounted for
* 366 - warning
* 404 - warning
* 442 - 9 annotations unaccounted for
* 484 - 5 annotations unaccounted for
* 608 - warning
* 815 - warning
* 863 - warning
* 865 - warning
* 942 - warning

**Comment by fedorov** [comment]
@notmatthancock I was investigating what is going on in the situations where some annotations were not assigned to any cluster, and I see something I cannot explain.

In 132 I see the following:

![image](https://user-images.githubusercontent.com/313942/48649245-0021ae00-e9c0-11e8-93a0-6193cbf306e8.png)

Note that the annotations denoted by the red arrow are all assigned to the same nodule cluster (nodule 4), while the annotations seem to be in a location that does not correspond to the actual nodule.

Annotations with the white arrow have not been assigned to any cluster, but seem to line up with the nodule in the image.

Any thoughts what might be going on?

I am now going to look if I see the same in the pylidc-generated visualizations, but I had troubles to get that one working in the past.

**Comment by fedorov** [comment]
It looks like somewhere somehow some annotations were flipped.

![image](https://user-images.githubusercontent.com/313942/48649492-bf766480-e9c0-11e8-957e-0ac7e9eb0568.png)
![image](https://user-images.githubusercontent.com/313942/48649514-d452f800-e9c0-11e8-8c05-cb06a4707948.png)

**Comment by fedorov** [comment]
Here's the code I used to debug this issue in pylidc proper:

```python
import matplotlib as mpl
mpl.rcParams['figure.dpi'] = 50
pid = 'LIDC-IDRI-0132'
scan = pl.query(pl.Scan).filter(pl.Scan.patient_id == pid).first()

annotations = pl.query(pl.Annotation).join(pl.Scan).filter(pl.Scan.patient_id == pid)
nodules = scan.cluster_annotations()

annotationsInNodulesList = []
for nCount,nodule in enumerate(nodules):
  print("  Nodule %d has %d annotations" % (nCount+1, len(nodule)))
  for a in nodule:
    annotationsInNodulesList.append(a.id)
if annotationsInNodules != annotations.count():
  print("   WARNING: %d annotations unaccounted for!" % (annotations.count()-annotationsInNodules))
  annotationsNotInNodules = []
  for a in annotations:
    if a.id not in annotationsInNodulesList:
      annotationsNotInNodules.append(a.id)
  print("IDs of annotations not in nodules: "+str(annotationsNotInNodules))

outliers = [ [i] for i in anns if i.id in annotationsNotInNodules]

scan.visualize(annotation_groups=outliers)
```

Looks like something is not right here, since the annotations arrows do not seem to point to anything in the image that looks like a nodule:

![image](https://user-images.githubusercontent.com/313942/48651478-1e3fdc00-e9c9-11e8-8f12-0ccca095eb5d.png)

I then tried to visualize the same nodule in pylidc as I do in Slicer, and it shows up in a different location than what I see in Slicer

```python
outliers = [ [i] for i in anns if i.id == 45] # _nodule_id 18366
scan.visualize(annotation_groups=outliers)
```

pylidc:

![image](https://user-images.githubusercontent.com/313942/48651657-e5eccd80-e9c9-11e8-9c4b-6f3e771dff45.png)

Slicer:

![image](https://user-images.githubusercontent.com/313942/48651668-f0a76280-e9c9-11e8-9fc0-1a8fb3977cea.png)

BUT if I then take a nodule that shows up lining up in Slicer, it does not seem to correspond to a nodule in pylidc!

 ```python
outliers = [ [i] for i in anns if i._nodule_id == "13112"]
scan.visualize(annotation_groups=outliers)
```

pylidc:

![image](https://user-images.githubusercontent.com/313942/48651706-28aea580-e9ca-11e8-9661-37b785e29c5f.png)

Slicer:

![image](https://user-images.githubusercontent.com/313942/48651713-319f7700-e9ca-11e8-82b2-3661cff96fa5.png)

@notmatthancock can you explain what is going on with those nodules not assigned to clusters, and why the seem to correspond to locations in the image where there are no nodules? Is this some geometry transformation issue? 

Once I understand what is going on the pylidc side, I will be in a better position to investigate what's going on in my conversion process.

**Comment by notmatthancock** [comment]
Thanks for all the details @fedorov. I'll look into this.

**Comment by fedorov** [comment]
I made a helper utility to examine cluster assignment and visualize specific annotations/groups of annotations here, if this helps: https://github.com/QIICR/lidc2dicom/blob/master/checkClusters.py

**Comment by fedorov** [comment]
I think I figured it out (unless I am missing something). All of those subjects that have annotations "unaccounted for" have more than 1 scan AND the number of annotations that have not been assigned to any clusters, which makes sense, since `cluster_annotations` is a function of `Scan`, so annotations from another scan are left out! Duh.

https://github.com/QIICR/lidc2dicom/blob/master/scansPerSubject.py

```
$ python scansPerSubject.py                                                                                          2.3.6
LIDC-IDRI-0132 has 2 scans
  Scan 1 has 12 annotations
  Scan 2 has 17 annotations
LIDC-IDRI-0151 has 2 scans
  Scan 1 has 3 annotations
  Scan 2 has 4 annotations
LIDC-IDRI-0315 has 2 scans
  Scan 1 has 18 annotations
  Scan 2 has 19 annotations
LIDC-IDRI-0332 has 2 scans
  Scan 1 has 8 annotations
  Scan 2 has 9 annotations
LIDC-IDRI-0355 has 2 scans
  Scan 1 has 1 annotations
  Scan 2 has 3 annotations
LIDC-IDRI-0365 has 2 scans
  Scan 1 has 4 annotations
  Scan 2 has 4 annotations
LIDC-IDRI-0442 has 2 scans
  Scan 1 has 10 annotations
  Scan 2 has 9 annotations
LIDC-IDRI-0484 has 2 scans
  Scan 1 has 5 annotations
  Scan 2 has 5 annotations
```

It was my fault not reading documentation. Sorry for bothering you with this!

**Comment by notmatthancock** [comment]
In your code, the line:
```
annotations = pl.query(pl.Annotation).join(pl.Scan).filter(pl.Scan.patient_id == pid)
```
retrieves all annotations for all scans having the given patient ID, which is not necessarily equivalent to retrieving all annotations for a given scan.

To retrieve all annotations for a given scan, we can simply do:
```
annotations = scan.annotations
```
which is syntactic sugar for the equivalent means of filtering on the foreign key:
```
annotations = pl.query(pl.Annotation).filter(pl.Annotation.scan_id == scan.id)
```
The `id` attribute is the unique primary key for every scan object while `patient_id` is not unique, which ties into your last comment. With the primary key, you can retrieve using the more concise syntax:
```
scan = query(Scan).get( scan_id )
```
with the drawback being of course that `scan_id` is just some arbitrary integer specific to pylidc.

The [`annotation_groups` argument of `scan.visualize`](https://github.com/pylidc/pylidc/blob/master/pylidc/Scan.py#L453-L457) function is intended to only accept the list of lists returned by calling `cluster_annotations` for the respective scan. In fact, this argument should really be a bool like `indicate_annotations=True` rather than accepting an arbitrary list in order to avoid potential funny business.

**Comment by fedorov** [comment]
Yes, thanks - I made the incorrect assumption initially that each subject has just one scan, and that's where it started.

**Comment by notmatthancock** [comment]
I see. Retrospectively a `Subject` class with a one to many relationship towards the `Scan` class would have made this more explicit.
